INSERT ... ON DUPLICATE KEY UPDATE ,Reporting the following error

sql:
insert into t_order (id, avl_inventory, `type`, 
create_person, create_time, update_person, update_time, is_deleted, version ) 
VALUES 
  (76667, 100, 1, 'xx(189870)', '2019-4-25 18:25:20', 'xx(189870)','2019-4-25 18:25:20', 0,0 ) , 
 (78143, 100, 1, 'xx(189870)', '2019-4-25 18:25:20',
 'xx(189870)', '2019-4-25 18:25:20', 0,0 )
 ON DUPLICATE KEY UPDATE avl_inventory = VALUES(avl_inventory),
 `type` = VALUES(`type`),
 create_time = VALUES(create_time), create_person = VALUES(create_person),
 update_time = VALUES(update_time), update_person = VALUES(update_person);

error:
Caused by: com.mysql.jdbc.exceptions.jdbc4.MySQLIntegrityConstraintViolationException: Duplicate entry '78143' for key 'PRIMARY'
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
	at com.mysql.jdbc.Util.handleNewInstance(Util.java:425)
	at com.mysql.jdbc.Util.getInstance(Util.java:408)
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:936)
	at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:3973)
	at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:3909)
	at com.mysql.jdbc.MysqlIO.sendCommand(MysqlIO.java:2527)
	at com.mysql.jdbc.ServerPreparedStatement.serverExecute(ServerPreparedStatement.java:1283)
	at com.mysql.jdbc.ServerPreparedStatement.executeInternal(ServerPreparedStatement.java:783)
	at com.mysql.jdbc.PreparedStatement.execute(PreparedStatement.java:1197)
	at com.zaxxer.hikari.pool.ProxyPreparedStatement.execute(ProxyPreparedStatement.java:44)
	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.execute(HikariProxyPreparedStatement.java)
	at org.apache.shardingsphere.shardingjdbc.executor.SQLExecuteCallbackFactory$2.executeSQL(SQLExecuteCallbackFactory.java:65)
	at org.apache.shardingsphere.shardingjdbc.executor.SQLExecuteCallbackFactory$2.executeSQL(SQLExecuteCallbackFactory.java:61)
	at org.apache.shardingsphere.core.execute.sql.execute.SQLExecuteCallback.execute0(SQLExecuteCallback.java:69)
	at org.apache.shardingsphere.core.execute.sql.execute.SQLExecuteCallback.execute(SQLExecuteCallback.java:58)

--- a/sharding-core/sharding-core-parse/sharding-core-parse-common/src/main/java/org/apache/shardingsphere/core/parse/antlr/extractor/impl/dml/insert/InsertColumnsExtractor.java
+++ b/sharding-core/sharding-core-parse/sharding-core-parse-common/src/main/java/org/apache/shardingsphere/core/parse/antlr/extractor/impl/dml/insert/InsertColumnsExtractor.java
@@ -43,8 +43,8 @@ public final class InsertColumnsExtractor implements OptionalSQLSegmentExtractor
     public Optional<InsertColumnsSegment> extract(final ParserRuleContext ancestorNode, final Map<ParserRuleContext, Integer> parameterMarkerIndexes) {
         Optional<ParserRuleContext> insertValuesClause = ExtractorUtils.findFirstChildNode(ancestorNode, RuleName.INSERT_VALUES_CLAUSE);
         return insertValuesClause.isPresent()
-                ? Optional.of(new InsertColumnsSegment(insertValuesClause.get().getStart().getStartIndex(), extractColumns(insertValuesClause.get(), parameterMarkerIndexes)))
-                : Optional.<InsertColumnsSegment>absent();
+                ? Optional.of(new InsertColumnsSegment(insertValuesClause.get().getStart().getStartIndex(), 
+                insertValuesClause.get().getStop().getStopIndex(), extractColumns(insertValuesClause.get(), parameterMarkerIndexes))) : Optional.<InsertColumnsSegment>absent();
     }
     
     private Collection<ColumnSegment> extractColumns(final ParserRuleContext ancestorNode, final Map<ParserRuleContext, Integer> parameterMarkerIndexes) {

--- a/sharding-core/sharding-core-parse/sharding-core-parse-common/src/main/java/org/apache/shardingsphere/core/parse/antlr/extractor/impl/dml/update/SetAssignmentsExtractor.java
+++ b/sharding-core/sharding-core-parse/sharding-core-parse-common/src/main/java/org/apache/shardingsphere/core/parse/antlr/extractor/impl/dml/update/SetAssignmentsExtractor.java
@@ -51,6 +51,6 @@ public final class SetAssignmentsExtractor implements OptionalSQLSegmentExtracto
                 assignmentSegments.add(assignmentSegment.get());
             }
         }
-        return Optional.of(new SetAssignmentsSegment(setAssignmentsClauseNode.get().getStart().getStartIndex(), assignmentSegments));
+        return Optional.of(new SetAssignmentsSegment(setAssignmentsClauseNode.get().getStart().getStartIndex(), setAssignmentsClauseNode.get().getStop().getStopIndex(), assignmentSegments));
     }
 }

--- a/sharding-core/sharding-core-parse/sharding-core-parse-common/src/main/java/org/apache/shardingsphere/core/parse/antlr/filler/common/dml/InsertColumnsFiller.java
+++ b/sharding-core/sharding-core-parse/sharding-core-parse-common/src/main/java/org/apache/shardingsphere/core/parse/antlr/filler/common/dml/InsertColumnsFiller.java
@@ -31,6 +31,7 @@ import org.apache.shardingsphere.core.parse.antlr.sql.token.InsertValuesToken;
  * Insert columns filler.
  *
  * @author zhangliang
+ * @author panjuan
  */
 @Setter
 public final class InsertColumnsFiller implements SQLSegmentFiller<InsertColumnsSegment>, ShardingTableMetaDataAwareFiller {
@@ -46,7 +47,7 @@ public final class InsertColumnsFiller implements SQLSegmentFiller<InsertColumns
             } else {
                 fillFromSQL(sqlSegment, insertStatement);
             }
-            insertStatement.getSQLTokens().add(new InsertValuesToken(sqlSegment.getStartIndex(), sqlStatement.getLogicSQL().length() - 1));
+            insertStatement.getSQLTokens().add(new InsertValuesToken(sqlSegment.getStartIndex(), sqlSegment.getStopIndex()));
         }
     }
     

--- a/sharding-core/sharding-core-parse/sharding-core-parse-common/src/main/java/org/apache/shardingsphere/core/parse/antlr/filler/encrypt/dml/EncryptSetAssignmentsFiller.java
+++ b/sharding-core/sharding-core-parse/sharding-core-parse-common/src/main/java/org/apache/shardingsphere/core/parse/antlr/filler/encrypt/dml/EncryptSetAssignmentsFiller.java
@@ -63,7 +63,7 @@ public final class EncryptSetAssignmentsFiller implements SQLSegmentFiller<SetAs
         InsertValue insertValue = getInsertValue(sqlSegment, insertStatement.getLogicSQL());
         insertStatement.getValues().add(insertValue);
         insertStatement.setParametersIndex(insertValue.getParametersCount());
-        insertStatement.getSQLTokens().add(new InsertSetToken(sqlSegment.getStartIndex(), insertStatement.getLogicSQL().length() - 1));
+        insertStatement.getSQLTokens().add(new InsertSetToken(sqlSegment.getStartIndex(), sqlSegment.getStopIndex()));
     }
     
     private InsertValue getInsertValue(final SetAssignmentsSegment sqlSegment, final String sql) {

--- a/sharding-core/sharding-core-parse/sharding-core-parse-common/src/main/java/org/apache/shardingsphere/core/parse/antlr/filler/sharding/dml/update/ShardingSetAssignmentsFiller.java
+++ b/sharding-core/sharding-core-parse/sharding-core-parse-common/src/main/java/org/apache/shardingsphere/core/parse/antlr/filler/sharding/dml/update/ShardingSetAssignmentsFiller.java
@@ -84,7 +84,7 @@ public final class ShardingSetAssignmentsFiller implements SQLSegmentFiller<SetA
         insertStatement.getValues().add(insertValue);
         insertStatement.getRouteConditions().getOrCondition().getAndConditions().add(andCondition);
         insertStatement.setParametersIndex(insertValue.getParametersCount());
-        insertStatement.getSQLTokens().add(new InsertSetToken(sqlSegment.getStartIndex(), insertStatement.getLogicSQL().length() - 1));
+        insertStatement.getSQLTokens().add(new InsertSetToken(sqlSegment.getStartIndex(), sqlSegment.getStopIndex()));
     }
     
     private int getColumnCountExcludeAssistedQueryColumns(final InsertStatement insertStatement) {

--- a/sharding-core/sharding-core-parse/sharding-core-parse-common/src/main/java/org/apache/shardingsphere/core/parse/antlr/sql/segment/dml/assignment/SetAssignmentsSegment.java
+++ b/sharding-core/sharding-core-parse/sharding-core-parse-common/src/main/java/org/apache/shardingsphere/core/parse/antlr/sql/segment/dml/assignment/SetAssignmentsSegment.java
@@ -27,6 +27,7 @@ import java.util.Collection;
  * Set assignments segment.
  *
  * @author zhangliang
+ * @author panjuan
  */
 @RequiredArgsConstructor
 @Getter

--- a/sharding-core/sharding-core-parse/sharding-core-parse-common/src/main/java/org/apache/shardingsphere/core/parse/antlr/sql/segment/dml/assignment/SetAssignmentsSegment.java
+++ b/sharding-core/sharding-core-parse/sharding-core-parse-common/src/main/java/org/apache/shardingsphere/core/parse/antlr/sql/segment/dml/assignment/SetAssignmentsSegment.java
@@ -34,5 +34,7 @@ public final class SetAssignmentsSegment implements SQLSegment {
     
     private final int startIndex;
     
+    private final int stopIndex;
+    
     private final Collection<AssignmentSegment> assignments;
 }

--- a/sharding-core/sharding-core-parse/sharding-core-parse-common/src/main/java/org/apache/shardingsphere/core/parse/antlr/sql/segment/dml/column/InsertColumnsSegment.java
+++ b/sharding-core/sharding-core-parse/sharding-core-parse-common/src/main/java/org/apache/shardingsphere/core/parse/antlr/sql/segment/dml/column/InsertColumnsSegment.java
@@ -34,5 +34,7 @@ public final class InsertColumnsSegment implements SQLSegment {
     
     private final int startIndex;
     
+    private final int stopIndex;
+    
     private final Collection<ColumnSegment> columns;
 }

--- a/sharding-core/sharding-core-parse/sharding-core-parse-common/src/main/java/org/apache/shardingsphere/core/parse/antlr/sql/segment/dml/column/InsertColumnsSegment.java
+++ b/sharding-core/sharding-core-parse/sharding-core-parse-common/src/main/java/org/apache/shardingsphere/core/parse/antlr/sql/segment/dml/column/InsertColumnsSegment.java
@@ -27,6 +27,7 @@ import java.util.Collection;
  * Insert columns segment.
  *
  * @author zhangliang
+ * @author panjuan
  */
 @RequiredArgsConstructor
 @Getter

--- a/sharding-core/sharding-core-parse/sharding-core-parse-common/src/main/java/org/apache/shardingsphere/core/parse/old/parser/clause/InsertSetClauseParser.java
+++ b/sharding-core/sharding-core-parse/sharding-core-parse-common/src/main/java/org/apache/shardingsphere/core/parse/old/parser/clause/InsertSetClauseParser.java
@@ -75,7 +75,7 @@ public abstract class InsertSetClauseParser implements SQLClauseParser {
         Optional<InsertValuesToken> insertValuesToken = insertStatement.findSQLToken(InsertValuesToken.class);
         Preconditions.checkState(insertValuesToken.isPresent());
         insertStatement.getSQLTokens().remove(insertValuesToken.get());
-        insertStatement.addSQLToken(new InsertSetToken(insertValuesToken.get().getStartIndex(), insertStatement.getLogicSQL().length() - 1));
+        insertStatement.addSQLToken(new InsertSetToken(insertValuesToken.get().getStartIndex(), insertValuesToken.get().getStopIndex()));
         do {
             SQLExpression left = basicExpressionParser.parse(insertStatement);
             Column column = null;

--- a/sharding-core/sharding-core-parse/sharding-core-parse-common/src/main/java/org/apache/shardingsphere/core/parse/old/parser/clause/InsertValuesClauseParser.java
+++ b/sharding-core/sharding-core-parse/sharding-core-parse-common/src/main/java/org/apache/shardingsphere/core/parse/old/parser/clause/InsertValuesClauseParser.java
@@ -90,7 +90,7 @@ public abstract class InsertValuesClauseParser implements SQLClauseParser {
         Optional<InsertValuesToken> insertValuesToken = insertStatement.findSQLToken(InsertValuesToken.class);
         Preconditions.checkState(insertValuesToken.isPresent());
         insertStatement.getSQLTokens().remove(insertValuesToken.get());
-        insertStatement.addSQLToken(new InsertValuesToken(insertValuesToken.get().getStartIndex(), insertStatement.getLogicSQL().length() - 1));
+        insertStatement.addSQLToken(new InsertValuesToken(insertValuesToken.get().getStartIndex(), insertValuesToken.get().getStopIndex()));
         do {
             lexerEngine.accept(Symbol.LEFT_PAREN);
             List<SQLExpression> sqlExpressions = new LinkedList<>();

--- a/sharding-core/sharding-core-parse/sharding-core-parse-common/src/main/java/org/apache/shardingsphere/core/parse/old/parser/clause/TableReferencesClauseParser.java
+++ b/sharding-core/sharding-core-parse/sharding-core-parse-common/src/main/java/org/apache/shardingsphere/core/parse/old/parser/clause/TableReferencesClauseParser.java
@@ -97,7 +97,7 @@ public class TableReferencesClauseParser implements SQLClauseParser {
         lexerEngine.nextToken();
         if (sqlStatement instanceof InsertStatement) {
             sqlStatement.addSQLToken(
-                    new InsertValuesToken(lexerEngine.getCurrentToken().getEndPosition() - lexerEngine.getCurrentToken().getLiterals().length(), sqlStatement.getLogicSQL().length() - 1));
+                    new InsertValuesToken(lexerEngine.getCurrentToken().getEndPosition() - lexerEngine.getCurrentToken().getLiterals().length(), lexerEngine.getCurrentToken().getEndPosition() - 1));
         }
         if (lexerEngine.skipIfEqual(Symbol.DOT)) {
             literals = lexerEngine.getCurrentToken().getLiterals();

--- a/sharding-core/sharding-core-rewrite/src/test/java/org/apache/shardingsphere/core/rewrite/SQLRewriteEngineTest.java
+++ b/sharding-core/sharding-core-rewrite/src/test/java/org/apache/shardingsphere/core/rewrite/SQLRewriteEngineTest.java
@@ -237,6 +237,26 @@ public final class SQLRewriteEngineTest {
     }
     
     @Test
+    public void assertRewriteForDuplicateKeyWithoutColumnsWithoutParameter() {
+        insertStatement.getColumnNames().add("name");
+        insertStatement.getColumnNames().add("id");
+        insertStatement.getTables().add(new Table("table_x", null));
+        insertStatement.addSQLToken(new TableToken(12, 20, "`table_x`", QuoteCharacter.BACK_QUOTE));
+        insertStatement.addSQLToken(new InsertValuesToken(21, 32));
+        InsertOptimizeResult insertOptimizeResult = new InsertOptimizeResult(InsertType.VALUES, Arrays.asList("name", "id"));
+        SQLExpression[] sqlExpressions = {new SQLNumberExpression(10), new SQLNumberExpression(1)};
+        insertOptimizeResult.addUnit(sqlExpressions, new Object[0], 0);
+        insertOptimizeResult.getUnits().get(0).getDataNodes().add(new DataNode("db0.table_1"));
+        TableUnit tableUnit = new TableUnit("db0");
+        tableUnit.getRoutingTables().add(new RoutingTable("table_x", "table_1"));
+        routeResult = new SQLRouteResult(insertStatement);
+        SQLRewriteEngine rewriteEngine = new SQLRewriteEngine(shardingRule, "INSERT INTO `table_x` VALUES (10) ON DUPLICATE KEY UPDATE name = VALUES(name)",
+                DatabaseType.MySQL, routeResult, Collections.emptyList(), new OptimizeResult(insertOptimizeResult));
+        assertThat(rewriteEngine.rewrite(false).toSQL(tableUnit, tableTokens, null, shardingDataSourceMetaData).getSql(), 
+                is("INSERT INTO `table_1` (name, id) VALUES (10, 1) ON DUPLICATE KEY UPDATE name = VALUES(name)"));
+    }
+    
+    @Test
     public void assertRewriteColumnWithoutColumnsWithoutParameter() {
         insertStatement.getColumnNames().add("name");
         insertStatement.getColumnNames().add("id");

--- a/sharding-core/sharding-core-rewrite/src/test/java/org/apache/shardingsphere/core/rewrite/SQLRewriteEngineTest.java
+++ b/sharding-core/sharding-core-rewrite/src/test/java/org/apache/shardingsphere/core/rewrite/SQLRewriteEngineTest.java
@@ -32,6 +32,7 @@ import org.apache.shardingsphere.core.parse.antlr.sql.statement.dml.SelectStatem
 import org.apache.shardingsphere.core.parse.antlr.sql.statement.dml.UpdateStatement;
 import org.apache.shardingsphere.core.parse.antlr.sql.token.EncryptColumnToken;
 import org.apache.shardingsphere.core.parse.antlr.sql.token.IndexToken;
+import org.apache.shardingsphere.core.parse.antlr.sql.token.InsertSetToken;
 import org.apache.shardingsphere.core.parse.antlr.sql.token.InsertValuesToken;
 import org.apache.shardingsphere.core.parse.antlr.sql.token.OffsetToken;
 import org.apache.shardingsphere.core.parse.antlr.sql.token.OrderByToken;
@@ -254,6 +255,26 @@ public final class SQLRewriteEngineTest {
                 DatabaseType.MySQL, routeResult, Collections.emptyList(), new OptimizeResult(insertOptimizeResult));
         assertThat(rewriteEngine.rewrite(false).toSQL(tableUnit, tableTokens, null, shardingDataSourceMetaData).getSql(), 
                 is("INSERT INTO `table_1` (name, id) VALUES (10, 1) ON DUPLICATE KEY UPDATE name = VALUES(name)"));
+    }
+    
+    @Test
+    public void assertRewriteForDuplicateKeyWithSetWithoutParameter() {
+        insertStatement.getColumnNames().add("name");
+        insertStatement.getColumnNames().add("id");
+        insertStatement.getTables().add(new Table("table_x", null));
+        insertStatement.addSQLToken(new TableToken(12, 20, "`table_x`", QuoteCharacter.BACK_QUOTE));
+        insertStatement.addSQLToken(new InsertSetToken(22, 43));
+        InsertOptimizeResult insertOptimizeResult = new InsertOptimizeResult(InsertType.SET, Arrays.asList("name", "id"));
+        SQLExpression[] sqlExpressions = {new SQLNumberExpression(10), new SQLNumberExpression(1)};
+        insertOptimizeResult.addUnit(sqlExpressions, new Object[0], 0);
+        insertOptimizeResult.getUnits().get(0).getDataNodes().add(new DataNode("db0.table_1"));
+        TableUnit tableUnit = new TableUnit("db0");
+        tableUnit.getRoutingTables().add(new RoutingTable("table_x", "table_1"));
+        routeResult = new SQLRouteResult(insertStatement);
+        SQLRewriteEngine rewriteEngine = new SQLRewriteEngine(shardingRule, "INSERT INTO `table_x` set name = 'a', id = 1 ON DUPLICATE KEY UPDATE name = VALUES(name)",
+                DatabaseType.MySQL, routeResult, Collections.emptyList(), new OptimizeResult(insertOptimizeResult));
+        assertThat(rewriteEngine.rewrite(false).toSQL(tableUnit, tableTokens, null, shardingDataSourceMetaData).getSql(),
+                is("INSERT INTO `table_1` SET name = 10, id = 1 ON DUPLICATE KEY UPDATE name = VALUES(name)"));
     }
     
     @Test

